utils/launch_redis.py

```diff
@@ -0,0 +1,257 @@
+        redisserver = "redis-server"
+    else:
+        redisserver = (
+            pathlib.Path(__file__).parent.parent
+            / "third-party/redis/src/redis-server"
+        ).resolve()
+    rediscli = "redis-cli" if is_cicd else os.path.dirname(redisserver) + "/redis-cli"
+    test_device = device if device is not None else os.environ.get(
+        "SMARTREDIS_TEST_DEVICE","cpu").lower()
+    if is_cicd:
+        redisai = os.getenv(f'REDISAI_{test_device.upper()}_INSTALL_PATH') + '/redisai.so'
+        redisai_modules = os.getenv("REDISAI_MODULES")
+        if redisai_modules is None:
+            raise RuntimeError("REDISAI_MODULES environment variable is not set!")
+        rai_clause = f"--loadmodule {redisai_modules}"
+    else:
+        if not rai_ver:
+            raise RuntimeError("RedisAI version not specified")
+        redisai_dir = (
+            pathlib.Path(__file__).parent.parent
+            / f"third-party/RedisAI/{rai_ver}/install-{test_device}"
+        ).resolve()
+        redisai = redisai_dir / "redisai.so"
+        tf_loc = redisai_dir / "backends/redisai_tensorflow/redisai_tensorflow.so"
+        torch_loc = redisai_dir / "backends/redisai_torch/redisai_torch.so"
+        rai_clause = f"--loadmodule {redisai} TF {tf_loc} TORCH {torch_loc}"
+    uds_clause = ""
+    if is_uds:
+        prepare_uds_socket(udsport)
+        uds_clause = f"--bind 127.0.0.1 --unixsocket {udsport} --unixsocketperm 777"
+    daemonize_clause = "--daemonize yes"
+    cluster_clause = "--cluster-enabled yes" if is_cluster else ""
+    prot_clause = "--protected-mode no" if is_cluster or is_uds else ""
+    save_clause = '--save ""' if is_cluster else ""
+
+    # Start servers
+    procs = []
+    for i in range(n_nodes):
+        l_port = port + i
+        port_clause = f"--port {l_port}" if not is_uds else "--port 0"
+        if is_cluster:
+            log_clause = f"--logfile {l_port}.log"
+            cluster_cfg_clause = f"--cluster-config-file {l_port}.conf"
+        else:
+            log_clause = "--logfile " + ("UDS.log" if is_uds else "single.log")
+            cluster_cfg_clause = ""
+        log_clause += " --loglevel notice"
+        cmd = f"{redisserver} {port_clause} {daemonize_clause} {cluster_clause} " + \
+              f"{cluster_cfg_clause} {log_clause} {uds_clause} {rai_clause} " + \
+              f"{prot_clause} {save_clause}"
+
+        print(cmd)
+        proc = Popen(cmd.split(), shell=False)
+        procs.append(proc)
+
+    # Make sure that all servers are up
+    # Let exceptions propagate to the caller
+    check_availability(n_nodes, port, udsport)
+    for proc in procs:
+        _ = proc.communicate(timeout=15)
+        if proc.returncode != 0:
+            raise RuntimeError("Failed to launch Redis server!")
+
+    # Create cluster for clustered Redis request
+    if n_nodes > 1:
+        cluster_str = ''
+        cluster_str += " ".join(f"127.0.0.1:{port + i}" for i in range(n_nodes))
```

Comment:
One last thing to remove an unnecessary assignment/strcpy 
```suggestion
        cluster_str = " ".join(f"127.0.0.1:{port + i}" for i in range(n_nodes))
```